File: API/1529_Volatility_Arbitrage_Spread_Oscillator_Model_VASOM/CS/VolatilityArbitrageSpreadOscillatorModelStrategy.cs
using System;
using System.Collections.Generic;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies;

/// <summary>
/// Trades VIX front month when the RSI of the spread between front and second month contracts is oversold.
/// Exits when RSI rises above the exit threshold.
/// </summary>
public class VolatilityArbitrageSpreadOscillatorModelStrategy : Strategy
{
	private readonly StrategyParam<int> _rsiPeriod;
	private readonly StrategyParam<int> _longThreshold;
	private readonly StrategyParam<int> _exitThreshold;
	private readonly StrategyParam<DataType> _candleType;
	private readonly StrategyParam<Security> _secondSecurity;

	private decimal _frontClose;
	private decimal _secondClose;
	private RelativeStrengthIndex _rsi;

	/// <summary>
	/// RSI period.
	/// </summary>
	public int RsiPeriod
	{
		get => _rsiPeriod.Value;
		set => _rsiPeriod.Value = value;
	}

	/// <summary>
	/// RSI level for long entry.
	/// </summary>
	public int LongThreshold
	{
		get => _longThreshold.Value;
		set => _longThreshold.Value = value;
	}

	/// <summary>
	/// RSI level for exit.
	/// </summary>
	public int ExitThreshold
	{
		get => _exitThreshold.Value;
		set => _exitThreshold.Value = value;
	}

	/// <summary>
	/// Candle type.
	/// </summary>
	public DataType CandleType
	{
		get => _candleType.Value;
		set => _candleType.Value = value;
	}

	/// <summary>
	/// Second month VIX future.
	/// </summary>
	public Security SecondSecurity
	{
		get => _secondSecurity.Value;
		set => _secondSecurity.Value = value;
	}

	public VolatilityArbitrageSpreadOscillatorModelStrategy()
	{
		_rsiPeriod = Param(nameof(RsiPeriod), 2)
			.SetGreaterThanZero()
			.SetDisplay("RSI Period", "Length of RSI", "Parameters");

		_longThreshold = Param(nameof(LongThreshold), 46)
			.SetRange(0, 100)
			.SetDisplay("Long Threshold", "RSI level to enter long", "Parameters");

		_exitThreshold = Param(nameof(ExitThreshold), 76)
			.SetRange(0, 100)
			.SetDisplay("Exit Threshold", "RSI level to exit", "Parameters");

		_candleType = Param(nameof(CandleType), TimeSpan.FromMinutes(5).TimeFrame())
			.SetDisplay("Candle Type", "Type of candles", "Parameters");

		_secondSecurity = Param(nameof(SecondSecurity), new Security { Id = "CBOE:VX2!" })
			.SetDisplay("Second Security", "Second month VIX future", "Parameters");
	}

	/// <inheritdoc />
	public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
		=> [(Security, CandleType), (SecondSecurity, CandleType)];

	/// <inheritdoc />
	protected override void OnReseted()
	{
		base.OnReseted();
		_frontClose = 0m;
		_secondClose = 0m;
		_rsi = null;
	}

	/// <inheritdoc />
	protected override void OnStarted(DateTimeOffset time)
	{
		base.OnStarted(time);

		_rsi = new RelativeStrengthIndex { Length = RsiPeriod };

		var frontSub = SubscribeCandles(CandleType);
		frontSub.Bind(ProcessFront).Start();

		var secondSub = SubscribeCandles(CandleType, security: SecondSecurity);
		secondSub.Bind(ProcessSecond).Start();

		var area = CreateChartArea();
		var rsiArea = CreateChartArea("Spread RSI");
		if (area != null)
		{
			DrawCandles(area, frontSub);
			DrawOwnTrades(area);
		}
		if (rsiArea != null)
			DrawIndicator(rsiArea, _rsi);
	}

	private void ProcessFront(ICandleMessage candle)
	{
		if (candle.State != CandleStates.Finished)
			return;

		_frontClose = candle.ClosePrice;
		TryTrade(candle);
	}

	private void ProcessSecond(ICandleMessage candle)
	{
		if (candle.State != CandleStates.Finished)
			return;

		_secondClose = candle.ClosePrice;
		TryTrade(candle);
	}

	private void TryTrade(ICandleMessage candle)
	{
		if (_frontClose == 0m || _secondClose == 0m)
			return;

		var spread = _frontClose - _secondClose;
		var rsiValue = _rsi.Process(spread, candle.OpenTime, true).ToDecimal();

		if (!_rsi.IsFormed || !IsFormedAndOnlineAndAllowTrading())
			return;

		if (Position <= 0 && rsiValue < LongThreshold)
			BuyMarket();
		else if (Position > 0 && rsiValue > ExitThreshold)
			SellMarket();
	}
}

File: API/1529_Volatility_Arbitrage_Spread_Oscillator_Model_VASOM/PY/volatility_arbitrage_spread_oscillator_model_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import RelativeStrengthIndex
from StockSharp.Algo.Strategies import Strategy


class volatility_arbitrage_spread_oscillator_model_strategy(Strategy):
    def __init__(self):
        super(volatility_arbitrage_spread_oscillator_model_strategy, self).__init__()
        self._rsi_period = self.Param("RsiPeriod", 14) \
            .SetDisplay("RSI Period", "Length of RSI", "Parameters")
        self._long_threshold = self.Param("LongThreshold", 35) \
            .SetDisplay("Long Threshold", "RSI level to enter long", "Parameters")
        self._exit_threshold = self.Param("ExitThreshold", 65) \
            .SetDisplay("Exit Threshold", "RSI level to exit", "Parameters")
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromMinutes(5))) \
            .SetDisplay("Candle Type", "Type of candles", "Parameters")
        self._front_close = 0.0
        self._second_close = 0.0
        self._rsi_val = 0.0
        self._prev_rsi = 0.0
        self._cooldown = 0

    @property
    def rsi_period(self):
        return self._rsi_period.Value

    @property
    def long_threshold(self):
        return self._long_threshold.Value

    @property
    def exit_threshold(self):
        return self._exit_threshold.Value

    @property
    def candle_type(self):
        return self._candle_type.Value

    def OnReseted(self):
        super(volatility_arbitrage_spread_oscillator_model_strategy, self).OnReseted()
        self._front_close = 0.0
        self._second_close = 0.0
        self._rsi_val = 0.0
        self._prev_rsi = 0.0
        self._cooldown = 0

    def OnStarted(self, time):
        super(volatility_arbitrage_spread_oscillator_model_strategy, self).OnStarted(time)
        rsi = RelativeStrengthIndex()
        rsi.Length = self.rsi_period
        subscription = self.SubscribeCandles(self.candle_type)
        subscription.Bind(rsi, self.on_process).Start()
        area = self.CreateChartArea()
        if area is not None:
            self.DrawCandles(area, subscription)
            self.DrawIndicator(area, rsi)
            self.DrawOwnTrades(area)

    def on_process(self, candle, rsi_value):
        if candle.State != CandleStates.Finished:
            return
        self._front_close = float(candle.ClosePrice)
        self._prev_rsi = self._rsi_val
        self._rsi_val = float(rsi_value)
        if self._cooldown > 0:
            self._cooldown -= 1
            return
        if self._prev_rsi == 0:
            return
        # Long entry when RSI crosses below threshold
        if self._rsi_val < self.long_threshold and self.Position <= 0:
            self.BuyMarket()
            self._cooldown = 45
        # Exit long / short entry when RSI crosses above threshold
        elif self._rsi_val > self.exit_threshold and self.Position >= 0:
            self.SellMarket()
            self._cooldown = 45

    def CreateClone(self):
        return volatility_arbitrage_spread_oscillator_model_strategy()
